main/2024/CG-06.md

```diff
@@ -93,10 +93,1859 @@ scheduled times if previous sessions end early.
+
+BT: I agree and think performance is important which is why I was asking. 
+
+AR: One note - problem to solve is constructing JS heap values in wasm. That’s the key interop problem.
+
+SD: Yes, you could see it as an interop problem between different wasm languages, too, at the GC level.
+
+AR: Right, and we’ve had many discussions about that and seamless interop magic was never the design goal because in our experience it never really works. It also collides with the goal of WebAssembly being a lower-level language. Any form of interop we would end up choosing would fit some language and not another. So far we have tried to keep it low-level with a couple exceptions like cast operators where we kind of went a little overboard. With all of the types here there is no hidden branching or hidden allocation and that is the line we need to not cross.
+
+SD: We would be fine with it being predictably slow! If i32 with that last bit could be heap allocated, sure, fine. We would prefer that for interop.
+
+AR: Isn’t it possible to express that right now in user space? In your own code you could say this is i31 and then the larger value right now.
+
+SD: Yes, that’s what we use the JS helper for. But we can’t do it entirely on the wasm side. We still need to ask JS to make a number for us.
+
+AR: Right but that is a separate problem. There you really need to create a JS interpretation and not really a boxed number. There is a hidden assumption on this last slide, in that these types would interop with JS the way you would expect them to.
+
+SD: Yes, and any other wasm language that chooses to use them.
+
+AR: My take would be that we could go a similar route with these that we are trying with string builtins. Particularly once we have something like type imports. Right now we use externref, but we could have interop modules in the JS API level that give you the primitives you need.
+
+SD: Yes that would work.
+
+CW: To echo Andreas, the direction I expect this to go would be to facilitate the allocation of JS objects instead of trying to make engines more clever behind the scenes.
+
+AR: One other comment re: closures: The main reason we removed them was concerns that we want a plain function pointer type, and don’t want to conflate it with a closure type. There might be room for closures as a future feature but would be a separate type. They are a richer thing with more expansive implementation.
+
+SD: Sure, that wouldn't fundamentally change how we use it.
+
+AR: With all of these things, it really takes someone to champion it and push it through. Now it seems that we do have a user.
+
+CW: To be fair, the rebuttal is often to have the engines implement these proposals as that often takes significant work.
+
+AR: Yes that too. 
+
+TL: Thank you very much!
+
+#### JS string builtins (Ryan Hunt)
+
+RH presenting. [TODO: slides]
```

Comment:
@eqrion do you have a slides link?

Reply:
Friendly bump @eqrion

Reply:
https://docs.google.com/presentation/d/111gkNGmzOtATU7daMKBjZ6mBqbIXF6jhd3e1542-0xo/edit#slide=id.g1ef40614c6f_0_21

Sorry, for the delay.